core/shared/src/main/scala/org/typelevel/log4cats/PagingSelfAwareStructuredLogger.scala

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2018 Christopher Davenport
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+import cats._
+import cats.implicits._
+
+import java.io.{ByteArrayOutputStream, PrintStream}
+import java.util.UUID
+
+object PagingSelfAwareStructuredLogger {
+
+  /**
+   * Wrap a SelfAwareStructuredLogger adding pagination functionality to avoid the truncation (hard splitting) by the
+   * underline logging provider. For the functions with a Throwable parameter, the stack trace is appended to the
+   * message before pagination.
+   *
+   * The actual text to be logged has 3 parts:
+   *
+   * Part 1. The chunk of message in each page;
+   * Part 2. This function will add the page header and footer;
+   * Part 3. The base class SelfAwareStructuredLogger may add tracing data etc logging context.
+   *
+   * The total of the above 3 parts should be smaller than the log size limit of the underline logging provider to
+   * avoid hard splitting.
+   *
+   * Example: Assume the log size limit of the underline logging provider is 75 Kb, setting pageSizeK to 64 leaves
+   * 11 Kb for (part 2) and (part 3).
+   *
+   * @param pageSizeK The size (in unit kibibyte) of the chunk of message in each page.
+   * @param maxPageNeeded The maximum number of pages to be logged.
+   * @param logger The SelfAwareStructuredLogger to be used to do the actual logging.
+   * @tparam F Effect type class.
+   * @return SelfAwareStructuredLogger with pagination.
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+  def withPaging[F[_]: Monad](pageSizeK: Int = 64, maxPageNeeded: Int = 999)(
+      logger: SelfAwareStructuredLogger[F]
+  ): SelfAwareStructuredLogger[F] =
+    new PagingSelfAwareStructuredLogger[F](pageSizeK, maxPageNeeded)(logger)
+
+  private class PagingSelfAwareStructuredLogger[F[_]: Monad](pageSizeK: Int, maxPageNeeded: Int)(
+      sl: SelfAwareStructuredLogger[F]
+  ) extends SelfAwareStructuredLogger[F] {
+    assert(pageSizeK > 0)
+    assert(maxPageNeeded > 0)
+
+    private val pageIndices = (1 to maxPageNeeded).toList
```

Comment:
I'm not sure whether `Range` has a `.traverse_` instance, but see if it compiles without the `.toList` to make it lazy.  If you can't, no big deal, but you're making a list of 999 numbers that you won't usually use.

Reply:
No luck. So I am creating a much smaller List of the size of numOfPages instead of 999 then taking from it.